File: 4.ts
function findMedianSortedArrays(nums1: number[], nums2: number[]): number {
    const m = nums1.length;
    const n = nums2.length;
    if (m > n) {
        return findMedianSortedArrays(nums2, nums1);
    }
    var iMin = 0, iMax = m;
    const k = (m + n + 1) >> 1;
    while (iMin < iMax) {
        var i: number = (iMin + iMax) >> 1;
        var j: number = k - i;
        if (nums1[i] < nums2[j - 1]) {
            iMin = i + 1
        }
        else {
            iMax = i
        }
    };
    i = iMin;
    j = k - i;
    const maxLeft = Math.max(i == 0 ? -Infinity : nums1[i - 1], j == 0 ? -Infinity : nums2[j - 1]);
    if ((m + n) % 2 === 1) {
        return maxLeft;
    }
    const minRight = Math.min(i == m ? Infinity : nums1[i], j == n ? Infinity : nums2[j]);
    return (maxLeft + minRight) / 2;
}

File: 4. Median of Two Sorted Arrays/4.go
func findMedianSortedArrays(nums1 []int, nums2 []int) float64 {
	m, n := len(nums1), len(nums2)
	if m > n {
		return findMedianSortedArrays(nums2, nums1)
	}
	l, r, half := 0, m, (m+n+1)>>1
	for l < r {
		i := (l + r) >> 1
		j := half - i
		if nums1[i] < nums2[j-1] {
			l = i + 1
		} else {
			r = i
		}
	}
	i := l
	j := half - i
	var maxLeft float64
	if i == 0 {
		maxLeft = float64(nums2[j-1])
	} else if j == 0 {
		maxLeft = float64(nums1[i-1])
	} else {
		maxLeft = float64(max(nums1[i-1], nums2[j-1]))
	}

	if (m+n)%2 == 1 {
		return maxLeft
	}
	var minRight float64
	if i == m {
		minRight = float64(nums2[j])
	} else if j == n {
		minRight = float64(nums1[i])
	} else {
		minRight = float64(min(nums1[i], nums2[j]))
	}
	return (maxLeft + minRight) / 2
}
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}